crates/sui-tool/src/commands.rs

```diff
@@ -61,48 +81,178 @@ fn make_clients(genesis: PathBuf) -> Result<BTreeMap<AuthorityName, NetworkAutho
+
+            for (version, resp) in versions {
                 writeln!(
                     f,
-                    "---- locked by : {}",
-                    lock.as_ref()
-                        .map(|l| format!("{:?}", l.digest()))
-                        .unwrap_or_else(|| "<not locked>".into())
+                    "-- version: {}",
+                    version.opt_debug("<version not available>")
                 )?;
+
+                match resp {
+                    Err(e) => writeln!(f, "Error fetching object: {}", e)?,
+                    Ok(resp) => {
+                        let objref = resp.requested_object_reference.opt_debug("<no object>");
+                        writeln!(f, "  -- ref: {}", objref)?;
+
+                        write!(f, "  -- cert:")?;
+                        match &resp.parent_certificate {
+                            None => writeln!(f, " <genesis>")?,
+                            Some(cert) => {
+                                let cert = format!("{}", cert);
+                                let cert = textwrap::indent(&cert, "     | ");
+                                write!(f, "\n{}", cert)?;
+                            }
+                        }
+
+                        if let Some(ObjectResponse { lock, .. }) = &resp.object_and_lock {
+                            // TODO maybe show object contents if we can do so meaningfully.
+                            writeln!(f, "  -- object: <data>")?;
+                            writeln!(f, "  -- locked by: {}", lock.opt_debug("<not locked>"))?;
+                        }
+                    }
+                }
             }
         }
         Ok(())
     }
 }
 
+async fn get_object(
+    client: &NetworkAuthorityClient,
+    id: ObjectID,
+    start_version: Option<u64>,
+    full_history: bool,
+) -> Vec<(Option<SequenceNumber>, Result<ObjectInfoResponse>)> {
+    let mut ret = Vec::new();
+    let mut version = start_version;
+
+    loop {
+        let resp = client
+            .handle_object_info_request(ObjectInfoRequest {
+                object_id: id,
+                request_kind: match version {
+                    None => ObjectInfoRequestKind::LatestObjectInfo(None),
+                    Some(v) => ObjectInfoRequestKind::PastObjectInfo(SequenceNumber::from_u64(v)),
+                },
+            })
+            .await
+            .map_err(anyhow::Error::from);
+
+        let resp_version = resp
+            .as_ref()
+            .ok()
+            .and_then(|r| r.requested_object_reference)
+            .map(|(_, v, _)| v.value());
+        ret.push((resp_version.map(SequenceNumber::from), resp));
+
+        version = match (version, resp_version) {
+            (Some(v), _) | (None, Some(v)) => {
+                if v == 0 || !full_history {
+                    break;
+                } else {
+                    Some(v - 1)
+                }
+            }
+            _ => break,
+        };
```

Comment:
The logic is a bit winding but here's what I'm thinking: if `full_history` is true, we from latest version or the specified version to version 0. The nullity of `resp_version` does not matter. Cuz logically an earlier version must exist before a later version (if we can't find it, then the authority's DB is probably corrupt). 

Simplified version:

```suggestion
        if !full_history || version == Some(0) {
            break;
        }
        version = Some(v-1);
```

Reply:
> Cuz logically an earlier version must exist before a later version (if we can't find it, then the authority's DB is probably corrupted).

That is incidentally true for now, but an authority can delete old object versions whenever it likes. We'll eventually have to do that to reclaim space. But even if it that wasn't allowed, a debugging tool like this should still make a best effort even if the authority is broken.